target/linux/generic/hack-4.4/950-net-patch-linux-kernel-to-support-shortcut-fe.patch

```diff
@@ -0,0 +1,145 @@
++#else
+ 	/* 2 or 4 bit hole */
++#endif
+ 
+ #ifdef CONFIG_NET_SCHED
+ 	__u16			tc_index;	/* traffic control index */
+diff --git a/net/Kconfig b/net/Kconfig
+index 6dedbb5..e8c6e6c 100644
+--- a/net/Kconfig
++++ b/net/Kconfig
+@@ -397,3 +397,6 @@ endif   # if NET
+ # Used by archs to tell that they support BPF_JIT
+ config HAVE_BPF_JIT
+ 	bool
++
++config SHORTCUT_FE
++	bool "Enables kernel network stack path for Shortcut  Forwarding Engine"
+diff --git a/net/core/dev.c b/net/core/dev.c
+index 11b9cda..3761691 100644
+--- a/net/core/dev.c
++++ b/net/core/dev.c
+@@ -2732,8 +2732,17 @@ static int xmit_one(struct sk_buff *skb, struct net_device *dev,
+ 	unsigned int len;
+ 	int rc;
+ 
++#ifdef CONFIG_SHORTCUT_FE
++	/* If this skb has been fast forwarded then we don't want it to
++	 * go to any taps (by definition we're trying to bypass them).
++	 */
++	if (!skb->fast_forwarded) {
++#endif
+ 	if (!list_empty(&ptype_all) || !list_empty(&dev->ptype_all))
+ 		dev_queue_xmit_nit(skb, dev);
++#ifdef CONFIG_SHORTCUT_FE
++	}
++#endif
+ 
+ #ifdef CONFIG_ETHERNET_PACKET_MANGLE
+ 	if (!dev->eth_mangle_tx ||
+@@ -3825,6 +3834,11 @@ void netdev_rx_handler_unregister(struct net_device *dev)
+ }
+ EXPORT_SYMBOL_GPL(netdev_rx_handler_unregister);
+ 
++#ifdef CONFIG_SHORTCUT_FE
++int (*fast_nat_recv)(struct sk_buff *skb) __rcu __read_mostly;
++EXPORT_SYMBOL_GPL(fast_nat_recv);
++#endif
++
+ /*
+  * Limit the use of PFMEMALLOC reserves to those protocols that implement
+  * the special handling of PFMEMALLOC skbs.
+@@ -3867,6 +3881,9 @@ static int __netif_receive_skb_core(struct sk_buff *skb, bool pfmemalloc)
+ 	bool deliver_exact = false;
+ 	int ret = NET_RX_DROP;
+ 	__be16 type;
++#ifdef CONFIG_SHORTCUT_FE
++	int (*fast_recv)(struct sk_buff *skb);
++#endif
+ 
+ 	net_timestamp_check(!netdev_tstamp_prequeue, skb);
+ 
+@@ -3893,6 +3910,16 @@ another_round:
+ 			goto out;
+ 	}
+ 
++#ifdef CONFIG_SHORTCUT_FE
++	fast_recv = rcu_dereference(fast_nat_recv);
++	if (fast_recv) {
++		if (fast_recv(skb)) {
++			ret = NET_RX_SUCCESS;
++			goto out;
++		}
++	}
++#endif
++
+ #ifdef CONFIG_NET_CLS_ACT
+ 	if (skb->tc_verd & TC_NCLS) {
+ 		skb->tc_verd = CLR_TC_NCLS(skb->tc_verd);
+diff --git a/net/netfilter/nf_conntrack_proto_tcp.c b/net/netfilter/nf_conntrack_proto_tcp.c
+index d41fdab..6f3cdfd 100644
+--- a/net/netfilter/nf_conntrack_proto_tcp.c
++++ b/net/netfilter/nf_conntrack_proto_tcp.c
+@@ -34,12 +34,22 @@
+ #include <net/netfilter/ipv6/nf_conntrack_ipv6.h>
+ 
+ /* Do not check the TCP window for incoming packets  */
++#ifdef CONFIG_SHORTCUT_FE
++int nf_ct_tcp_no_window_check __read_mostly = 0;
++EXPORT_SYMBOL_GPL(nf_ct_tcp_no_window_check);
++#else
+ static int nf_ct_tcp_no_window_check __read_mostly = 1;
++#endif
+ 
+ /* "Be conservative in what you do,
+     be liberal in what you accept from others."
+     If it's non-zero, we mark only out of window RST segments as INVALID. */
++#ifdef CONFIG_SHORTCUT_FE
++int nf_ct_tcp_be_liberal __read_mostly;
```

Comment:
Why is this needed, care to explain ?
If it was not done by you do you have the origin of that patch ?

Reply:
This symbol along with tcp_no_window_check is used in both connection managers and trigger some additional checks if un/set.
It also originates from qsdk SFE repo, placed in patches folder.

Reply:
Actually I’ve just copy/pasted patches for k4.4 from QSDK. As you see I already have these points cleaned out in patch for k4.9. I’m actually not sure if there is a need for k4.4 here, as LEDE is switching off it in master branch and unlikely this package makes its way into stable 17.01

Reply:
ops, wrong point

Reply:
Thanks!

Reply:
Dear dissent1
Could you tell me what 's the different betweent yours and glw's 
who's respositoryof fast patch is more faster?
thank you

Reply:
@jack338c I'm not dissent1, but from what I can see between the two, @dissent1 has stayed closer to the upstream project in QSDK whereas gwlim's patchset includes a heap of compile-time optimization for the mips 24 and 74k arch.

Reply:
@philjohn 
thanks for patience and help
I test both with my netgear wndr4300 74k and found gwlim's is more faster,especially on playing videos ,I can see the buffer is more quicker on gwlim's
but I also want to use dissent1,it's simple for compling

Reply:
hi @philjohn 
I just bought a R7800
if I want compile a firmware from https://github.com/dissent1/r7800/tree/bd2a683e28367e850f4f83f50c81ff9a73c92af6

how to git clone the repository with shortcut-fe
if I git clone https://github.com/dissent1/r7800.git ,there will be no shortcut-fe.
please help me ,thank you

Reply:
add .patch to the end of this URL, so the end is 1269.patch, then apply it to the source tree.

Failing that, have a look into hnyman's builds in the forum as I believe one of his now has fastpath support.

Reply:
sorry what is "this URL"?
Could I just download a zipfile from https://github.com/dissent1/r7800/tree/bd2a683e28367e850f4f83f50c81ff9a73c92af6 ,then exact it for compile?
thanks for fast reply